### PR TITLE
Function for creating a populated Pw builder, from an existing input file

### PR DESCRIPTION
**Pull request by chrisjsewell**
`aiida_quantumespresso.tools.pwinputparser.create_builder_from_file` is essentially code,
extracted and updated, from the deprecated `PwimmigrantCalculation.create_input_nodes` method.

The advantage of having it as an isolated function, is that it can be used for either:

- Setting up a calculation for submission, or
- The first step in creating a fully immigrated run (with an existing output file)

### CODE REVIEW

File: aiida_quantumespresso/tools/pwinputparser.py [hunk 1]
```diff
@@ -318,3 +321,96 @@ def parse_k_points(txt):
...
+    :type pseudo_folder_path: aiida.common.folders.Folder or str or None
+    :param use_first: parsed to UpfData.get_or_create
+    :type use_first: bool
+    :raises NotImplementedError: if the structure is not ibrav=0
+    :return: a builder instance for PwCalculation
+
+    """
+    pw_calc_cls = CalculationFactory('quantumespresso.pw')
+
+    inputs = {}
```
Comment (sphuber) [inline aiida_quantumespresso/tools/pwinputparser.py]: Why not straight away construct a builder here? `builder = PwCalculation.get_builder()` and attach the inputs as you parse them

File: aiida_quantumespresso/tools/pwinputparser.py [hunk 2]
```diff
@@ -318,3 +321,96 @@ def parse_k_points(txt):
...
+    inputs["metadata"] = metadata
```
Comment (sphuber) [inline aiida_quantumespresso/tools/pwinputparser.py]: Minor nitpick, I am trying to maintain the standard of using single quotes for identifier strings, such as keys in dictionaries

File: tests/tools/test_immigrate.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    in_folderpath = os.path.abspath(in_foldername)
+
+    upf_foldername = os.path.join('tests', 'fixtures', 'pseudos')
+    upf_folderpath = os.path.abspath(upf_foldername)
+    si_upf = generate_upf_data('Si')
+    si_upf.store()
+
+    builder = create_builder_from_file(in_folderpath, 'test_pw_default.in', code, metadata, upf_folderpath)
+
+    assert builder["code"] == code
```
Comment (sphuber) [inline tests/tools/test_immigrate.py]: Same here, please use single quotes for identifier strings

**Comment by chrisjsewell** [comment]
> Just some minor changes. Other than that all good. Thanks @chrisjsewell !

No problem. FYI I've also written a function for creating immigrated runs

```python
from aiida.common.datastructures import CalcJobState
from aiida.common.folders import Folder, SandboxFolder
from aiida.common.links import LinkType
from aiida.engine import ProcessState
from aiida.engine.daemon.execmanager import parse_results
from aiida.engine.utils import instantiate_process
from aiida.manage.manager import get_manager
from aiida.orm import FolderData, RemoteData
from aiida.plugins import CalculationFactory

def immigrate_run(builder, folder, infile, outfile, remote_path, xml_file=None):
    """ """

    runner = get_manager().get_runner()
    pw_calc_cls = CalculationFactory('quantumespresso.pw')
    process = instantiate_process(runner, pw_calc_cls, **builder)
    calc_node = process.node
    calc_node.set_attribute('immigrated', True)

    # create and populate retrieved folder
    retrieved_files = FolderData()
    retrieved_files.put_object_from_file(
        folder.get_abs_path(infile), process.metadata.options.input_filename)
    retrieved_files.put_object_from_file(
        folder.get_abs_path(outfile), process.metadata.options.output_filename)
    if xml_file is not None:
        retrieved_files.put_object_from_file(
            folder.get_abs_path(xml_file), 'data-file-schema.xml')    
    retrieved_files.add_incoming(calc_node, link_type=LinkType.CREATE,
                                 link_label=calc_node.link_label_retrieved)
    retrieved_files.store()

    # create remote data folder
    calc_node.set_remote_workdir(remote_path)
    remotedata = RemoteData(computer=calc_node.computer, remote_path=remote_path)
    remotedata.add_incoming(calc_node, link_type=LinkType.CREATE,
                            link_label='remote_folder')
    remotedata.store()

    # parse output and link outgoing nodes
    calc_node.set_state(CalcJobState.PARSING)
    with SandboxFolder() as temp_retrieved:
        exit_code = process.parse(temp_retrieved.abspath)
    process.update_outputs()

    # finalise calc node
    calc_node.delete_state()
    calc_node.delete_checkpoint()
    calc_node.set_process_state(ProcessState.FINISHED)
    calc_node.set_exit_status(exit_code.status)
    calc_node.set_exit_message(exit_code.message)
    calc_node.seal()

    return calc_node
```

I think it does a decent job, of replicating the main data.
Or do you think there is a better way?

**Pull request merged**